go-controller/pkg/ovn/ovn.go

```diff
@@ -59,23 +61,63 @@ type ACLLoggingLevels struct {
 	Deny  string `json:"deny,omitempty"`
 }
 
-// Controller structure is the object which holds the controls for starting
-// and reacting upon the watched resources (e.g. pods, endpoints)
-type Controller struct {
+// GenericController structure is place holder for all fields shared among controllers.
+type GenericController struct {
 	client       clientset.Interface
 	kube         kube.Interface
 	watchFactory *factory.WatchFactory
-	stopChan     <-chan struct{}
+	podRecorder  metrics.PodRecorder
 
-	// FIXME DUAL-STACK -  Make IP Allocators more dual-stack friendly
-	masterSubnetAllocator        *subnetallocator.SubnetAllocator
-	hybridOverlaySubnetAllocator *subnetallocator.SubnetAllocator
+	// event recorder used to post events to k8s
+	recorder record.EventRecorder
+
+	// libovsdb northbound client interface
+	nbClient libovsdbclient.Client
+
+	// libovsdb southbound client interface
+	sbClient libovsdbclient.Client
 
+	// has SCTP support
 	SCTPSupport bool
+}
+
+// ControllerManager structure is the object manages all controllers for all networks
+type ControllerManager struct {
+	GenericController
+
+	// default wait group and stop channel, used by default network controller
+	defaultWg       *sync.WaitGroup
+	defaultStopChan <-chan struct{}
+
+	// unique identity for controllerManager running on different ovnkube-master instance,
+	// used for leader election
+	identity string
+
+	// default network controller
+	defaultController *Controller
+
+	// controller for all networks, key is netName of net-attach-def, value is *Controller
+	allOvnControllers sync.Map
```

Comment:
Why a `sync.Map`? is there going to be more than 1 thread in the worker pool for NADs?

Reply:
there is no worker pool as you might think of. the creation/deletion of controllers are processed in net-attach-def adding/delete watch handlers, which are serialized. Except the default controller though, which is done when ovnkube-master starts.

Reply:
But what I am thinking though is that we should be doing a level-driven controller here, with it's own independent work queue (like the service controller). That's something which we plan do for the other controllers in 4.14 so I I think it would not make sense to do it here in the old event driven controller fashion and then having to re-code that again in 4.14.

Reply:
@trozet thoughts on this?

Reply:
Ideally yeah we want this to be level driven with a workqueue since that is what we are asking everyone else to do when they add a new controller.